### PR TITLE
Add reflection-based Search index definition

### PR DESCRIPTION
**Pull request by idg10**
This implements https://github.com/Azure/azure-sdk-for-net/issues/2427

For the unsquashed history see https://github.com/idg10/azure-sdk-for-net/tree/reflection-schema which includes discussion with @brjohnstmsft as he reviewed my work on this feature.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@idg10__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by azurecla** [comment]
__@idg10__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

File: src/Search/Microsoft.Azure.Search/Customizations/Indexes/FieldBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                    {
+                        field.Analyzer = AnalyzerName.Create(analyzerAttribute.Name);
+                    }
+                    else if ((searchAnalyzerAttribute = attribute as SearchAnalyzerAttribute) != null)
+                    {
+                        field.SearchAnalyzer = AnalyzerName.Create(searchAnalyzerAttribute.Name);
+                    }
+                    else if ((indexAnalyzerAttribute = attribute as IndexAnalyzerAttribute) != null)
+                    {
+                        field.IndexAnalyzer = AnalyzerName.Create(((IndexAnalyzerAttribute) attribute).Name);
```
Comment (brjohnstmsft) [inline src/Search/Microsoft.Azure.Search/Customizations/Indexes/FieldBuilder.cs]: I missed this earlier. It looks like the cast to `IndexAnalyzerAttribute` is unnecessary if you use the `indexAnalyzerAttribute` variable instead.

File: src/Search/Microsoft.Azure.Search/Customizations/Indexes/Models/AnalyzerName.cs [hunk 2]
```diff
@@ -466,31 +466,515 @@ public sealed class AnalyzerName : ExtensibleEnum<AnalyzerName>
...
+            /// <summary>
+            /// Standard Lucene analyzer.
+            /// </summary>
+            public const string StandardLucene = "standard.lucene";
+
+            /// <summary>
+            /// Standard ASCII Folding Lucene analyzer.
+            /// <see href="https://msdn.microsoft.com/library/azure/mt605304.aspx#Analyzers" /> 
+            /// </summary>
+            public static readonly AnalyzerName StandardAsciiFoldingLucene =
```
Comment (brjohnstmsft) [inline src/Search/Microsoft.Azure.Search/Customizations/Indexes/Models/AnalyzerName.cs]: This should be a string.

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: src/Search/Microsoft.Azure.Search/Customizations/Indexes/FieldBuilder.cs (hunk 1 above)
Comment (idg10) [inline src/Search/Microsoft.Azure.Search/Customizations/Indexes/FieldBuilder.cs]: Fixed in https://github.com/Azure/azure-sdk-for-net/pull/2536/commits/f46c44cde80ec93d73b15c292ff31f1ca72d531c (resquashed)

File: src/Search/Microsoft.Azure.Search/Customizations/Indexes/Models/AnalyzerName.cs (hunk 2 above)
Comment (idg10) [inline src/Search/Microsoft.Azure.Search/Customizations/Indexes/Models/AnalyzerName.cs]: Fixed in https://github.com/Azure/azure-sdk-for-net/pull/2536/commits/f46c44cde80ec93d73b15c292ff31f1ca72d531c (resquashed)

**Comment by idg10** [comment]
I've fixed both of the problems you found (the unnecessary repeated cast, and my bizarre error in `AnalyzerName.AsString.StandardAsciiFoldingLucene`). I presumed you still want this to be a single commit in the event that you accept the PR, so I've squashed it again and force pushed to my PR branch.

**Review by bitsforthought — APPROVED** [review]

**Comment by markcowl** [comment]
@azuresdkci add to whitelist

**Comment by markcowl** [comment]
LGT^M once the build passes

**Comment by brjohnstmsft** [comment]
@markcowl The build passed; Please merge

**Pull request merged**